Add PERT distribution

dhardy opened the pull request:
Fixes #778 @barabadzhi can you review?

Testing is incomplete, as with most distributions. We need to add code for histogram/PDF testing.

Error types may see additional tweaking before the next release, but I think this will do for now @vks?

barabadzhi commented on rand_distr/src/triangular.rs:
```diff
@@ -62,37 +62,165 @@ impl Distribution<f64> for Triangular {
+    
+    /// Set up the PERT distribution with defined `min`, `max`, `mode` and
+    /// `shape`.
+    pub fn new_shape(min: f64, max: f64, mode: f64, shape: f64) -> Result<Pert, PertError> {
```
:+1: for parametrized constructor

barabadzhi commented on rand_distr/src/triangular.rs:
```diff
@@ -62,37 +62,165 @@ impl Distribution<f64> for Triangular {
+            (5., 25., 25.),
+        ] {
+            let _distr = Pert::new(min, max, mode).unwrap();
+            // TODO: test correctness
```
Here is the correct test example I used:
```rust
#[test]
fn test_pert_sample() {
    let mut rng = crate::test::rng(860);
    let distr = Pert::new(45., 225., 90.).unwrap();
    let seq = distr.sample(&mut rng);
    let expected = 106.34517425256126;

    assert!(seq == expected);
}
```

Rough hand calculations as follows:
```
optimistic: 45
pessimistic: 225
most likely: 90

estimate average: (45+225+90)/3 = 120
estimate PERT: (45+225+4×90)/6 = 105
```

dhardy replied:
I think though it's pure luck that the sample lies close to your estimate? This is a random number distribution driven by random input. Note that the triangular distribution is tested by forcing the RNG to output "half", however interpretation is dependent on distribution implementations (which for this one includes the Beta and Gamma distrs).

barabadzhi replied:
Hmm, not really.. :thinking: 
![image](https://user-images.githubusercontent.com/8407846/56497897-bb1a1e80-64ff-11e9-9e34-af72e8ab81ba.png)

This results can be proven by Monte Carlo simulation:
![image](https://user-images.githubusercontent.com/8407846/56497872-a047aa00-64ff-11e9-839d-5e89a0b2858a.png)

Thus your `PERT` implementation is good enough for most cases, IMHO. :slightly_smiling_face:

dhardy replied:
This still doesn't test correctness. The results from the included test vector are also quite close (μ = 4, var = 12/7), so I don't think including your example is useful.

barabadzhi replied:
@dhardy I never asked for inclusion. :confused:  
Just decried my own experience checking out and testing the branch. :slightly_smiling_face:

dhardy commented:
Rebased and reorganised commits, moved Pert distr to a new file, renamed `Pert::new_shape` → `Pert::new_with_shape`; renamed `Dirichlet::new_with_param` → `Dirichlet::new_with_size` to match.
